### PR TITLE
Improved JSON functionality and added features

### PR DESCRIPTION
**Pull request by SmilyOrg**

### CODE REVIEW

File: docs/examples/JSONExample/src/JSONExample.ls [hunk 1]
```diff
@@ -90,11 +123,51 @@ package
...
             sprite.y = 120;
             stage.addChild(sprite);
+            
+            
+            var json = new JSON();
+            
+            json.initObject();
+            
+            
+            // Circular reference tests
```
Comment (bengarney) [inline docs/examples/JSONExample/src/JSONExample.ls]: Failing test?

File: loom/script/native/core/system/lmJSON.cpp [hunk 2]
```diff
@@ -645,5 +723,5 @@ static int registerSystemJSON(lua_State *L)
 
 void installSystemJSON()
 {
-    NativeInterface::registerNativeType<JSON>(registerSystemJSON);
+    NativeInterface::registerManagedNativeType<JSON>(registerSystemJSON);
```
Comment (bengarney) [inline loom/script/native/core/system/lmJSON.cpp]: Why this change?

File: sdk/src/system/JSON.ls [hunk 3]
```diff
@@ -52,7 +54,268 @@ enum JSONType
...
+     * @return  The JSON object parsed from the string.
+     */
+    public static function parse(json:String):JSON {
+        var j = new JSON();
+        Debug.assert(j.loadString(json), "JSON failed to load");
+        return j;
+    }
+    
+    /**
+     * Traverses through the object's fields and builds a JSON tree
```
Comment (bengarney) [inline sdk/src/system/JSON.ls]: Is a JSON tree like a JSON string?

**Comment by bengarney** [comment]
Nitted!

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (SmilyOrg) [inline docs/examples/JSONExample/src/JSONExample.ls]: Hmm, it's failing because the example is set to `latest` and not `dev`.

File: loom/script/native/core/system/lmJSON.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/script/native/core/system/lmJSON.cpp]: Otherwise I get:
```Managed mismatch for type system.JSON, script declaration specifies native while native bindings are
 unmanaged```

**Comment by SmilyOrg** [comment]
Counterpoint!

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (bengarney) [inline docs/examples/JSONExample/src/JSONExample.ls]: Oh OK. So we should reinstate it when on a newer version of Loom? Is there a way we can put an if around it to automatically enable it on "good" versions of Loom?

File: loom/script/native/core/system/lmJSON.cpp (hunk 2 above)
Comment (bengarney) [inline loom/script/native/core/system/lmJSON.cpp]: Right but why did you change both sides? That error is if you change only one.

**Comment by bengarney** [comment]
Countercounterpoint!

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (SmilyOrg) [inline docs/examples/JSONExample/src/JSONExample.ls]: Wait a second, shouldn't the test / build server automatically set the version to dev while testing? So I'm not sure why the test would fail? Seems green too.

File: loom/script/native/core/system/lmJSON.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/script/native/core/system/lmJSON.cpp]: I think that some of the LoomScript-only functions I added necessitated the switch to a managed type.

**Comment by SmilyOrg** [comment]
Etc.

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (bengarney) [inline docs/examples/JSONExample/src/JSONExample.ls]: It should. Does it not? Can you update the Rakefile if not?

File: loom/script/native/core/system/lmJSON.cpp (hunk 2 above)
Comment (bengarney) [inline loom/script/native/core/system/lmJSON.cpp]: What happens if you back it out on both sides?

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (SmilyOrg) [inline docs/examples/JSONExample/src/JSONExample.ls]: I don't see it failing though? What failed?

File: loom/script/native/core/system/lmJSON.cpp (hunk 2 above)
Comment (SmilyOrg) [inline loom/script/native/core/system/lmJSON.cpp]: Wouldn't backing out on the script side mean removing the script-side LS-only methods? Removing the entire point of the pull? :P

It's probable I'm not entirely understanding what a manage native type entails.

Comment (bengarney) [inline loom/script/native/core/system/lmJSON.cpp]: Ahh OK. That's why. Gotcha. I'm fine with this.

File: docs/examples/JSONExample/src/JSONExample.ls (hunk 1 above)
Comment (bengarney) [inline docs/examples/JSONExample/src/JSONExample.ls]: I want to include the commented out bit!

Comment (SmilyOrg) [inline docs/examples/JSONExample/src/JSONExample.ls]: Ohhhhh! That makes more sense, I'll look into it when I get the chance :)

Comment (SmilyOrg) [inline docs/examples/JSONExample/src/JSONExample.ls]: Ah, the commented out bit trips an assertion (equivalent to Flash throwing an error). I don't know a good way to handle that in an example, so I just left it commented and added a line explaining it.

**Comment by SmilyOrg** [comment]
Ok, addressed the last nit, hopefully.
